### PR TITLE
Support for having LighthouseJobs and Tekton jobs in multiple namespaces

### PR DESCRIPTION
**Pull request by keskad**
Part of https://github.com/jenkins-x/lighthouse/pull/1424

More context and related changes are there: https://github.com/jenkins-x-charts/jxboot-helmfile-resources/pull/425 and in main PR in lighthouse (https://github.com/jenkins-x/lighthouse/pull/1424)

All changes there are done in context of parent PR (namespaces support, multitenacy)

Release notes
--------------------

### Little steps to parametrize namespace selection in all JX places

- `jx gitops webhook update` have additional commandline switch `--dev-ns` to be able to switch to other namespace (a little step for making Jenkins X not aligned to only one namespace)

### Cluster scoped pods clean up to support running jobs in multiple namespaces

- `jx gitops gc pods` is able to search for Jenkins X namespaces labelled by bootjob with the environment name and then execute pods clean up only in those namespaces (required `RoleBinding` per namespace - this is done in separate repo where helms are - define a list of namespaces in `jxJobsNamespaces` in your `jx-global-values.yaml` to propagate bindings)

### Fix that enables multi-namespaced Helm releases inside bootrepo helmfiles

- `jx gitops namespace --dir-mode --dir config-root/namespaces` is fixed to not enforce namespace when `.metadata.namespace` is already defined in yaml file. This potentially broken some helms that had cross-namespace resources such as RoleBindings. Instead of changing namespace in a resource, the resource is moved to directory named same as `.metadata.namespace`

- `jx gitops helmfile move` is fixed to not enforce namespace when `.metadata.namespace` is already defined in yaml file. This potentially broken some helms that had cross-namespace resources such as RoleBindings.

### CODE REVIEW

**Comment by keskad** [comment]
Added a linked PR: https://github.com/jenkins-x/jx-helpers/pull/362

**Comment by keskad** [comment]
Added a linked PR: https://github.com/jenkins-x/jx-helpers/pull/362

File: go.mod [hunk 1]
```diff
@@ -20,6 +20,7 @@ require (
 	github.com/jenkins-x/jx-kube-client/v3 v3.0.2
 	github.com/jenkins-x/jx-logging/v3 v3.0.6
 	github.com/jenkins-x/lighthouse-client v0.0.295
+	github.com/keskad/jx-helpers/v3 v3.0.67-0.20220312071418-f6abb1196975  // todo: revert before merging and bump to latest version
```
Comment (keskad) [inline go.mod:23]: Before merging please merge https://github.com/jenkins-x/jx-helpers/pull/362 , then switch there back to JX repository and bump a version.

File: pkg/cmd/namespace/namespace.go [hunk 2]
```diff
@@ -5,24 +5,28 @@ import (
...
 	"github.com/jenkins-x/jx-helpers/v3/pkg/cobras/helper"
 	"github.com/jenkins-x/jx-helpers/v3/pkg/cobras/templates"
 	"github.com/jenkins-x/jx-helpers/v3/pkg/files"
-	"github.com/jenkins-x/jx-helpers/v3/pkg/kyamls"
 	"github.com/jenkins-x/jx-helpers/v3/pkg/options"
 	"github.com/jenkins-x/jx-helpers/v3/pkg/stringhelpers"
 	"github.com/jenkins-x/jx-helpers/v3/pkg/termcolor"
 	"github.com/jenkins-x/jx-helpers/v3/pkg/yamls"
 	"github.com/jenkins-x/jx-logging/v3/pkg/log"
+	"github.com/keskad/jx-helpers/v3/pkg/kyamls" // todo: revert before merging
```
Comment (keskad) [inline pkg/cmd/namespace/namespace.go:19]: Before merging please merge https://github.com/jenkins-x/jx-helpers/pull/362 , then switch there back to JX repository and bump a version.

File: pkg/cmd/namespace/namespace_test.go [hunk 3]
```diff
@@ -1,16 +1,16 @@
...
 	"os"
 	"path/filepath"
 	"strings"
 	"testing"
 
 	"github.com/google/go-cmp/cmp"
-	"github.com/jenkins-x-plugins/jx-gitops/pkg/cmd/namespace"
 	"github.com/jenkins-x/jx-helpers/v3/pkg/files"
-	"github.com/jenkins-x/jx-helpers/v3/pkg/kyamls"
+	"github.com/keskad/jx-helpers/v3/pkg/kyamls" // todo: revert before merging
```
Comment (keskad) [inline pkg/cmd/namespace/namespace_test.go:13]: Before merging please merge https://github.com/jenkins-x/jx-helpers/pull/362 , then switch there back to JX repository and bump a version.

**Comment by keskad** [comment]
Possibly related
---------------------

The fixes to `jx gitops namespace` and `jx helmfile move` are possibly fixing https://github.com/jenkins-x-charts/jxboot-helmfile-resources/pull/412

I noticed that it resolved an issue with `gc-jobs` job not being created in proper namespace, so the bootjobs `kind: Job` were not cleaned up.

The resource is defined following:

```
apiVersion: batch/v1beta1
kind: CronJob
metadata:
  name: jx-gcjobs
  namespace: {{ .Values.gc.jobs.cronjob.ns | default "jx-git-operator" | quote }}
```

And because of the bug it was always placed in `jx` instead of `jx-git-operator`, now I can see that it was fixed by the way:

```
 7 files changed, 27 insertions(+), 5 deletions(-)
 create mode 100644 config-root/cluster/namespaces/jx-git-operator.yaml
 create mode 100644 config-root/cluster/namespaces/jx-test1.yaml
 rename config-root/namespaces/{jx => jx-git-operator}/jxboot-helmfile-resources/gcjobs-rb.yaml (93%)
 rename config-root/namespaces/{jx => jx-git-operator}/jxboot-helmfile-resources/gcjobs-role.yaml (92%)
 rename config-root/namespaces/{jx => jx-git-operator}/jxboot-helmfile-resources/jx-gcjobs-cronjob.yaml (97%)
 rename config-root/namespaces/{jx => jx-git-operator}/jxboot-helmfile-resources/jx-gcjobs-sa.yaml (89%)
 rename config-root/namespaces/{jx => jx-test1}/jxboot-helmfile-resources/gcpods-jx-test1-rb.yaml (96%)
```

So the `gc-jobs` was placed where it was planned to.

**Comment by keskad** [comment]
Ready for review :smile: 
@yelhouti, @ankitm123? :slightly_smiling_face:

File: pkg/cmd/helmfile/move/move_test.go [hunk 4]
```diff
@@ -1,12 +1,12 @@
-package move_test
+package move
```
Comment (ankitm123) [inline pkg/cmd/helmfile/move/move_test.go:1]: why this change? We want to only test exported functions.

File: pkg/cmd/namespace/namespace_test.go [hunk 5]
```diff
@@ -1,14 +1,14 @@
-package namespace_test
+package namespace
```
Comment (ankitm123) [inline pkg/cmd/namespace/namespace_test.go:1]: same, keep it as namespace_test.

**Review by ankitm123 — COMMENTED** [review]
took a quick look, will review more tonight :)

File: pkg/cmd/namespace/namespace_test.go (hunk 5 above)
Comment (keskad) [reply pkg/cmd/namespace/namespace_test.go:1]: Sure.

File: pkg/cmd/helmfile/move/move_test.go (hunk 4 above)
Comment (keskad) [reply pkg/cmd/helmfile/move/move_test.go]: Fixed.

File: pkg/cmd/gc/pods/pods.go [hunk 6]
```diff
@@ -121,3 +157,28 @@ func (o *Options) MatchesPod(pod *corev1.Pod) (bool, time.Duration) {
...
 }
+
+// findJXNamespaces looks for namespace names that are owned by Jenkins X
+func (o *Options) findJXNamespaces(ctx context.Context) ([]string, error) {
+	var matched []string
+	namespaces, err := o.KubeClient.CoreV1().Namespaces().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return []string{}, errors.Wrap(err, "cannot list namespaces to look for pods")
+	}
+	for _, n := range namespaces.Items { //nolint:gocritic
```
Comment (ankitm123) [inline pkg/cmd/gc/pods/pods.go:168]: why nolint? Does it give false positive? or is it because of `_, n` instead of using `for k := range ...`

Comment (ankitm123) [inline pkg/cmd/gc/pods/pods.go:168]: why nolint? Does it give false positive? or is it because of `_, n` instead of using `for k := range ...`

File: pkg/cmd/gc/pods/pods_test.go [hunk 7]
```diff
@@ -0,0 +1,61 @@
+package pods
```
Comment (ankitm123) [inline pkg/cmd/gc/pods/pods_test.go:1]: `pods_test` for package name.

Comment (ankitm123) [inline pkg/cmd/gc/pods/pods_test.go:1]: `pods_test` for package name.

File: pkg/cmd/namespace/namespace.go [hunk 8]
```diff
@@ -181,25 +191,105 @@ func (o *Options) lazyCreateNamespaceResource(ns string) error {
 }
 
 // UpdateNamespaceInYamlFiles updates the namespace in yaml files
-func UpdateNamespaceInYamlFiles(dir string, ns string, filter kyamls.Filter) error { //nolint:gocritic
-	modifyFn := func(node *yaml.RNode, path string) (bool, error) {
+func UpdateNamespaceInYamlFiles(rootDir string, dir string, ns string, filter kyamls.Filter, shouldMoveFiles bool) ([]string, error) { //nolint:gocritic
```
Comment (ankitm123) [inline pkg/cmd/namespace/namespace.go:194]: `rootDir, dir, ns string` should fix the gocritic linter warning i think 🤔

Comment (ankitm123) [inline pkg/cmd/namespace/namespace.go:194]: `rootDir, dir, ns string` should fix the gocritic linter warning i think 🤔

File: pkg/cmd/gc/activities/activities.go [hunk 9]
```diff
@@ -108,6 +108,8 @@ func NewCmdGCActivities() (*cobra.Command, *Options) {
 	cmd.Flags().DurationVarP(&o.ReleaseAgeLimit, "release-age", "r", time.Hour*24*30, "Maximum age to keep PipelineActivities for Releases")
 	cmd.Flags().DurationVarP(&o.PipelineRunAgeLimit, "pipelinerun-age", "", time.Hour*12, "Maximum age to keep completed PipelineRuns for all pipelines")
 	cmd.Flags().DurationVarP(&o.ProwJobAgeLimit, "prowjob-age", "", time.Hour*24*7, "Maximum age to keep completed ProwJobs for all pipelines")
+	cmd.Flags().BoolVarP(&o.AllNamespaces, "all-namespaces", "", false, "Work in all namespaces (requires a ClusterRole)")
```
Comment (ankitm123) [inline pkg/cmd/gc/activities/activities.go:111]: need to run `make docs` to update the documentation.

Comment (ankitm123) [inline pkg/cmd/gc/activities/activities.go:111]: need to run `make docs` to update the documentation.

File: pkg/cmd/helmfile/move/move.go [hunk 10]
```diff
@@ -311,3 +315,21 @@ func (o *Options) moveFilesToClusterOrNamespacesFolder(dir, ns, releaseName, cha
 	}
 	return nil
 }
+
+// BuildOutputDirPath Builds output directory path. Returns: directory path where file should be placed
+func (o *Options) BuildOutputDirPath(node *yaml.RNode, path, ns, pathName string) string {
+	return filepath.Join(o.NamespacesDir, GetTargetNamespace(node, path, ns), pathName)
+}
+
+func GetTargetNamespace(node *yaml.RNode, path, ns string) string {
```
Comment (ankitm123) [inline pkg/cmd/helmfile/move/move.go:324]: Will this be used outside of jx-gitops? If not we dont need to export these, unless u have it this way to add tests for these functions ...

Comment (ankitm123) [inline pkg/cmd/helmfile/move/move.go:324]: Will this be used outside of jx-gitops? If not we dont need to export these, unless u have it this way to add tests for these functions ...

File: pkg/cmd/gc/pods/pods.go (hunk 6 above)
Comment (keskad) [reply pkg/cmd/gc/pods/pods.go:168]: Do you have a suggestion? It says loop iteration = ~400 bytes and I should use a pointer I don't have an idea how to do this, Kubernetes librarry does not return a pointer.

File: pkg/cmd/helmfile/move/move.go (hunk 10 above)
Comment (keskad) [reply pkg/cmd/helmfile/move/move.go:324]: I didn't wanted to export it, but you requrested to set package name to `move_test` and I was unable to test it if it was not exported.

Comment (keskad) [reply pkg/cmd/helmfile/move/move.go:324]: Same here as https://github.com/jenkins-x-plugins/jx-gitops/pull/835#discussion_r826643000

File: pkg/cmd/gc/activities/activities.go (hunk 9 above)
Comment (keskad) [reply pkg/cmd/gc/activities/activities.go:111]: Sure will do

Comment (keskad) [reply pkg/cmd/gc/activities/activities.go:111]: Sure, will do

File: pkg/cmd/namespace/namespace.go (hunk 8 above)
Comment (keskad) [reply pkg/cmd/namespace/namespace.go:194]: Yeah, I guess my IDE appended that nolint thing :thinking:

File: pkg/cmd/gc/pods/pods_test.go (hunk 7 above)
Comment (keskad) [reply pkg/cmd/gc/pods/pods_test.go:1]: I would not be able to test internal methods if I do so. What do you think?

Comment (keskad) [reply pkg/cmd/gc/pods/pods_test.go:1]: Same as https://github.com/jenkins-x-plugins/jx-gitops/pull/835#discussion_r826644017

File: pkg/cmd/gc/pods/pods.go (hunk 6 above)
Comment (ankitm123) [reply pkg/cmd/gc/pods/pods.go:168]:
```
for k := range namespaces.Items {
  n := namespaces.Items[k]
  labels := n.GetLabels()
```

File: pkg/cmd/gc/pods/pods_test.go (hunk 7 above)
Comment (ankitm123) [reply pkg/cmd/gc/pods/pods_test.go:1]: we should not be testing internal methods, only test exported functions (there are exceptions to the rules though). This will force us to think about the external API.

File: pkg/cmd/gc/pods/pods.go (hunk 6 above)
Comment (keskad) [reply pkg/cmd/gc/pods/pods.go:168]: Thanks, will try it :grin:

File: pkg/cmd/gc/pods/pods_test.go (hunk 7 above)
Comment (keskad) [reply pkg/cmd/gc/pods/pods_test.go:1]: Hmm... but then tests will become more functional tests (testing more facades than single units of code, less precise) than unit testing. But if it is ok for you then I can change.

Comment (keskad) [reply pkg/cmd/gc/pods/pods_test.go:1]: Ok, I will try to test just exported methods.

Comment (keskad) [reply pkg/cmd/gc/pods/pods_test.go]: Fixed.

File: pkg/cmd/gc/pods/pods.go (hunk 6 above)
Comment (keskad) [reply pkg/cmd/gc/pods/pods.go]: Thanks, fixed.

Comment (keskad) [reply pkg/cmd/gc/pods/pods.go]: Fixed.

File: pkg/cmd/helmfile/move/move.go (hunk 10 above)
Comment (keskad) [reply pkg/cmd/helmfile/move/move.go]: Fixed.

Comment (keskad) [reply pkg/cmd/helmfile/move/move.go]: I don't think it would be used outside. Fixed.

**Comment by keskad** [comment]
I just commited fixes, @ankitm123 :slightly_smiling_face:

File: pkg/cmd/webhook/update/update.go [hunk 11]
```diff
@@ -88,6 +89,7 @@ func NewCmdWebHookVerify() (*cobra.Command, *Options) {
 	cmd.Flags().StringVarP(&o.HMAC, "hmac", "", "", "Don't use the HMAC token from the cluster, use the provided token")
 	cmd.Flags().StringVarP(&o.Endpoint, "endpoint", "", "", "Don't use the endpoint from the cluster, use the provided endpoint")
 	cmd.Flags().BoolVarP(&o.WarnOnFail, "warn-on-fail", "", false, "If enabled lets just log a warning that we could not update the webhook")
+	cmd.Flags().StringVarP(&o.EnforceNamespace, "enforce-namespace", "", "", "Disable namespace auto detection, instead use specified namespace")
```
Comment (ankitm123) [inline pkg/cmd/webhook/update/update.go:92]: can be confusing, how about &DevNamespace? 
`StringVarP(&o.DevNamespace, "dev-ns",`
This argument is ultimately setting/getting the dev namespace. wdyt?

**Review by ankitm123 — COMMENTED** [review]
minor comments, overall looks great!

File: pkg/cmd/gc/pods/pods.go [hunk 12]
```diff
@@ -121,3 +174,29 @@ func (o *Options) MatchesPod(pod *corev1.Pod) (bool, time.Duration) {
...
+		labels := n.GetLabels()
+
+		if _, okEnv := labels[JXEnvironmentLabel]; okEnv {
+			if _, okGitOps := labels[JXGitOpsLabel]; okGitOps {
+				log.Logger().Infof("Found Jenkins X namespace '%s' by environment labels", n.Name)
+				matched = append(matched, n.Name)
+			}
+		}
+
+		if n.Name == "jx-git-operator" {
```
[the remark this answers is not in the document]
Comment (keskad) [reply pkg/cmd/gc/pods/pods.go:196]: Sure.

File: pkg/cmd/gc/pods/pods.go [hunk 13]
```diff
@@ -121,3 +174,29 @@ func (o *Options) MatchesPod(pod *corev1.Pod) (bool, time.Duration) {
 	}
 	return age > o.Age, age
 }
+
+// findJXNamespaces looks for namespace names that are owned by Jenkins X
```
[the remark this answers is not in the document]
Comment (keskad) [reply pkg/cmd/gc/pods/pods.go:178]: Happy to read :)

File: pkg/cmd/gc/pods/pods.go [hunk 14]
```diff
@@ -51,54 +58,100 @@ func NewCmdGCPods() (*cobra.Command, *Options) {
...
+
+		// run in single, selected namespace
+		namespaces = []string{o.Namespace}
+	} else {
+		log.Logger().Info("Looking for JX namespaces")
+
+		// run in all namespaces that are owned by Jenkins X
+		namespaces, err = o.findJXNamespaces(ctx)
+		if err != nil {
+			return []string{}, errors.Wrap(err, "cannot perform garbage collection of pods")
```
[the remark this answers is not in the document]
Comment (keskad) [reply pkg/cmd/gc/pods/pods.go:104]: Sure

File: pkg/cmd/webhook/update/update.go (hunk 11 above)
Comment (keskad) [reply pkg/cmd/webhook/update/update.go:92]: Yup. The autodetection mechanism is taking the namespace from the `team` label of the namespace as I remember. It seems also that the concept of `team` was in JX v2.

Comment (keskad) [reply pkg/cmd/webhook/update/update.go:92]: Will change to dev-ns.

File: docs/cmd/jx-gitops_webhook_update.md [hunk 15]
```diff
@@ -28,6 +28,7 @@ Updates the webhooks for all the source repositories optionally filtering by own
 ```
   -b, --batch-mode                 Runs in batch mode without prompting for user input
       --endpoint string            Don't use the endpoint from the cluster, use the provided endpoint
+      --enforce-namespace string   Disable namespace auto detection, instead use specified namespace
```
Comment (ankitm123) [inline docs/cmd/jx-gitops_webhook_update.md:31]: Need to re-run `make docs`

**Comment by keskad** [comment]
Fixed code review issues, rebased to properly name commits so the chanelog would be nice, added release changelog.

One thing - I accidently touched msvticket's commit description, and it have changed commit id now, sorry for that, don't know how to revert that.

File: docs/cmd/jx-gitops_webhook_update.md (hunk 15 above)
Comment (keskad) [reply docs/cmd/jx-gitops_webhook_update.md:31]: Ups :grin:

Comment (keskad) [reply docs/cmd/jx-gitops_webhook_update.md]: Done

**Comment by ankitm123** [comment]
> Fixed code review issues, rebased to properly name commits so the chanelog would be nice, added release changelog.
> 
> One thing - I accidently touched msvticket's commit description, and it have changed commit id now, sorry for that, don't know how to revert that.

use git reflog, may be?

**Comment by keskad** [comment]
> > Fixed code review issues, rebased to properly name commits so the chanelog would be nice, added release changelog.
> > One thing - I accidently touched msvticket's commit description, and it have changed commit id now, sorry for that, don't know how to revert that.
> 
> use git reflog, may be?

OK, I made a new local branch from original repo, cherry-picked my commits and pushed there.

**Comment by keskad** [comment]
Possibly related: https://github.com/jenkins-x-plugins/jx-gitops/pull/514 (may be closed?)

**Comment by nicolaerosia** [comment]
does this fix https://github.com/jenkins-x-charts/jxboot-helmfile-resources/pull/412 ?

**Comment by yelhouti** [comment]
Yes it is supposed to

**Comment by keskad** [comment]
@nicolaerosia Yes, but also linked PR is needed to resolve it: https://github.com/jenkins-x-charts/jxboot-helmfile-resources/pull/425/files#diff-6f5f287a8a9a6c3fd8cfd6eaa22fb0ea4b466240f2bcc6f507d233d11ce90df8R13

Without that change in RoleBinding it will get access denied still. So just a little moment more is needed in the review process I think :grin:
